### PR TITLE
MAINT: Fix the code style to our C-Style-Guide

### PR DESCRIPTION
**Pull request by fengluoqiuwu**
Fix the code style to our C-Style-Guide. More precisely, mainly do the following things:
 - **Rename Misleading Parameter Names**:
Rename parameters that could be misleading. For example, in line 590, the parameter `n_chars` in the function `buffer_memcpy` could be misunderstood as referring to the number of characters.
 - **Rename Methods and Structures**:
Rename some methods and structs to make sure that they are fixed to our c style conduct. (such as line 1722)

### CODE REVIEW

File: numpy/_core/src/umath/string_buffer.h [hunk 1]
```diff
@@ -24,14 +24,14 @@ enum class ENCODING {
...
     ISDIGIT,
     ISSPACE,
     ISALNUM,
     ISLOWER,
     ISUPPER,
     ISTITLE,
     ISNUMERIC,
+    ISDECIMAL,
     STR_LEN,
 };
```
Comment (charris) [inline numpy/_core/src/umath/string_buffer.h:37]: It was in alphabetic order, not by length.

**Comment by charris** [comment]
I'm going to leave this to @ngoldbaum. It isn't  that it is wrong, but changing variable names is mostly churn. I can see the removal of `void` arguments in what is C++ code, likewise using `false` to initialize C++ bool type, but wholesale renames are a bit much. Programmers will differ in what they regard as suitable names.

**Comment by fengluoqiuwu** [comment]
I've reordered it alphabetically. Thanks for the heads-up!

**Comment by fengluoqiuwu** [comment]
> I'm going to leave this to @ngoldbaum. It isn't that it is wrong, but changing variable names is mostly churn. I can see the removal of `void` arguments in what is C++ code, likewise using `false` to initialize C++ bool type, but wholesale renames are a bit much. Programmers will differ in what they regard as suitable names.

For some names, I simply added an underscore (`_`) to align with our C-style guide. For names with more substantial changes, I felt the original names could be somewhat misleading. If any relevant developers disagree with my changes, I’m more than willing to revert them or changing into a better name.

**Review by ngoldbaum — COMMENTED** [review]
I think the changes to function arguments are OK.

Please revert all the changes to function names and enums. IMO it's all unnecessary churn.

If anyone happens to have a PR open touching this code they'll have to handle merge conflicts and carefully tease out exactly what changed. That's just one example of the unnecessary work changes like this cause.

We're all human and it's very hard to follow a style guide that's not enforced.

IMO it would be a better use of your time to come up with some sort of clang-format config that gets used in CI on PR diffs to ensure we're following a style guide for any new code. More radically, you might also argue for mass-reformatting all the C and C++ code to follow a uniform style enforced via clang-format and then enforce it thereafter in CI for all code, but I doubt that you'll get developer consensus to do that.

**Comment by charris** [comment]
We do have a `.clang-format` file. It is a bit messy because it enforces include file order. It could also use a longer line length. Linux now enforces a 100 character line length with 80 characters preferred, see https://www.theregister.com/2020/06/01/linux_5_7/.

**Comment by ngoldbaum** [comment]
Sure, we have one, but it's not enforced by CI, so it's not really very useful for contributors or for ensuring that the code style remains uniform.

**Comment by seberg** [comment]
Is the C style fuzzy like the Python one (where black type style flies, but only PEP 8 violations matter?).
Because if it isn't, I am not sure I want it in CI, unless we just auto(re)format the whole repo.  I don't want to keep iterating on random formatting that is only enforced in the parts that are touched.

**Comment by ngoldbaum** [comment]
[`clang-format-diff`](https://github.com/llvm-mirror/clang/blob/master/tools/clang-format/clang-format-diff.py) is a thing. We'd need to experiment with it probably.

I didn't want to presume anything for the project by suggesting to work on automatic formatting. More that if someone took some time and carefully argued how to introduce it in a non-disruptive way, I think that would be more useful than needing to manually review a bunch of changes to enforce the style guide in existing code.

**Comment by fengluoqiuwu** [comment]
> Please revert all the changes to function names and enums. IMO it's all unnecessary churn.

I’ve reverted all changes to the function names and enums as requested.

File: numpy/_core/src/umath/string_buffer.h [hunk 2]
```diff
@@ -895,10 +895,12 @@ string_find(Buffer<enc> buf1, Buffer<enc> buf2, npy_int64 start, npy_int64 end)
     npy_intp pos;
     switch(enc) {
         case ENCODING::UTF8:
-            pos = fastsearch(start_loc, end_loc - start_loc, buf2.buf, buf2.after - buf2.buf, -1, FAST_SEARCH);
+            pos = fastsearch(start_loc, end_loc - start_loc, buf2.buf,
+                    buf2.after - buf2.buf, -1, FAST_SEARCH);
```
Comment (charris) [inline numpy/_core/src/umath/string_buffer.h:899]: Strictly speaking, this line should line up with the arguments above. Same below.

Comment (charris) [reply numpy/_core/src/umath/string_buffer.h:899]: Running .clang-format on this file will probably fix that up, and maybe do some extra stuff you may want to undo.

Comment (fengluoqiuwu) [reply numpy/_core/src/umath/string_buffer.h:899]: Uhhhh, running the `.clang-format` can sometimes lead to changes that violate the C-style guide, such as altering the formatting of all methods within a struct :
```c++
// origin code
inline Buffer()
{
    buf = after = NULL;
}
// after running .clang-format
inline Buffer() { buf = after = NULL; }
```
To be honest, I'm unsure how to configure `.clang-format` to align with our C-style guide.

Comment (charris) [reply numpy/_core/src/umath/string_buffer.h:899]: clang-format is sort of the replacement for the C-style guide.

File: numpy/_core/src/umath/string_buffer.h [hunk 3]
```diff
@@ -1325,15 +1333,16 @@ string_lrstrip_chars(Buffer<enc> buf1, Buffer<enc> buf2, Buffer<enc> out, STRIPT
 template <typename char_type>
 static inline npy_intp
 findslice_for_replace(CheckedIndexer<char_type> buf1, npy_intp len1,
-                      CheckedIndexer<char_type> buf2, npy_intp len2)
+                       CheckedIndexer<char_type> buf2, npy_intp len2)
```
Comment (charris) [inline numpy/_core/src/umath/string_buffer.h:1336]: This was correct before the change.

File: numpy/_core/src/umath/string_buffer.h (hunk 2 above)
Comment (charris) [reply numpy/_core/src/umath/string_buffer.h:899]: In case you are wondering, the indentation should look like this:
```
            pos = fastsearch(start_loc, end_loc - start_loc, buf2.buf,
                             buf2.after - buf2.buf, -1, FAST_SEARCH);
```
The other indentation is only when there are no arguments included in the first line. I have made exceptions to that, but mostly for exception messages. I'm not trying to be difficult here, but if you are working on style you need to become familiar with the conventions.

Comment (fengluoqiuwu) [reply numpy/_core/src/umath/string_buffer.h:899]: > I'm not trying to be difficult here

I know, I appreciate your feedback, both of our goal is to enhance code readability. 
My change stems from my own difficulties in reading the code, and I hope to help others avoid similar issues.

> In case you are wondering, the indentation should look like this:

But according to the [C-Style Guide](https://numpy.org/neps/nep-0045-c_style_guide.html), it says that "It isn’t necessary to line everything up with the opening parenthesis of the function call." 
image

And in some cases (maybe the function's or param's name is too long or the amount of retract is a bit large), align the first parameter of each row could lead to several line break, so I agree with our C-Style Guide.

File: numpy/_core/src/umath/string_buffer.h (hunk 3 above)
Comment (fengluoqiuwu) [reply numpy/_core/src/umath/string_buffer.h]: Thank you for pointing that out; I hadn't noticed it before.

**Comment by seberg** [comment]
@ngoldbaum if you like the changes now, do you want to just merge it?
